### PR TITLE
Explain vm/i

### PR DESCRIPTION
**Pull request by xpivarc**
**What this PR does / why we need it**:
PR is adding validation to CRDs. This will allow users of kubevirt to use standard tool kubectl to show descriptions/doc of these objects. 

This also enables validation of our CRs by Kubernetes API.
 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

We need to review the usage of markers(such as `+optional` `+nullabe`) so we don't get denials.

Few things need to be changed by provided tools as controller-gen is not handling it correctly for now.

**Release note**:
```release-note
Added support of `kubectl explain` for Kubevirt resources.
```

### CODE REVIEW

**Comment by rmohr** [comment]
Hi @xpivarc thanks for the PR.

I think what we really want is switching from our generator to `https://github.com/kubernetes/kube-openapi/tree/master/cmd/openapi-gen`.

Once you have `openapi-gen` created you  can do something like:

```
cd staging/src/kubevirt.io/client-go/api/v1/
penapi-gen --alsologtostderr -i ./ -o ./staging/src/kubevirt.io/client-go/api/v1/ -p blub -h ~/go/src/kubevirt.io/kubevirt/hack/boilerplate/boilerplate.go.txt
```

and it will generate the openapi specification instead of our mechanism.

**Comment by enp0s3** [comment]
@xpivarc Also, if you are still working on the PR you can add [WIP] to the PR subject

File: pkg/virt-operator/install-strategy/strategy.go [hunk 1]
```diff
@@ -218,15 +223,26 @@ func dumpInstallStrategyToBytes(strategy *InstallStrategy) []byte {
 	return b.Bytes()
 }
 
+func patchValidation(crd *extv1beta1.CustomResourceDefinition, name string) *extv1beta1.CustomResourceDefinition {
+	if name != "vmi" && name != "vm" {
```
Comment (davidvossel) [inline pkg/virt-operator/install-strategy/strategy.go:227]: why filter out non vmi/vm crds here? We want all our objects to have a schema.

File: pkg/virt-operator/install-strategy/validations_generated.go [hunk 2]
```diff
@@ -0,0 +1,7068 @@
+package installstrategy
+
+var resources map[string]string = map[string]string{
```
Comment (davidvossel) [inline pkg/virt-operator/install-strategy/validations_generated.go:3]: I think this generated file is actually kind of clever. It lets us maintain the ability to programatically build the CRDs the way we are now, yet still use the standard controller-gen logic to generate the schema on our behave. 

basically, it lets us go with the flow of how crd generation logic is being supported in the k8s ecosystem while still maintaining our current design patterns.

File: staging/src/kubevirt.io/client-go/Makefile [hunk 3]
```diff
@@ -0,0 +1,5 @@
+include $(addprefix ../../../../vendor/github.com/openshift/build-machinery-go/make/, \
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/Makefile:1]: why do we need a makefile to accomplish this?

File: tools/crd-validation-generator/validation-generator.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	filepath := fmt.Sprintf("%svalidations_generated.go", outputDir)
+	os.Remove(filepath)
+	file, err := os.OpenFile(filepath, os.O_CREATE|os.O_WRONLY, 0644)
+	defer file.Close()
+	if err != nil {
+		panic(fmt.Errorf("Failed to create go file %v, %v", filepath, err))
+	}
+	// w := bufio.NewWriter(file)
+	file.WriteString("package installstrategy\n\n")
+	file.WriteString("var resources map[string]string = map[string]string{\n")
```
Comment (davidvossel) [inline tools/crd-validation-generator/validation-generator.go:55]: can we call this map something more descriptive than `resources`

File: pkg/virt-operator/install-strategy/strategy.go [hunk 5]
```diff
@@ -218,15 +223,26 @@ func dumpInstallStrategyToBytes(strategy *InstallStrategy) []byte {
...
 func GenerateCurrentInstallStrategy(config *operatorutil.KubeVirtDeploymentConfig, addMonitorServiceResources bool, operatorNamespace string) (*InstallStrategy, error) {
 
 	strategy := &InstallStrategy{}
 
-	strategy.crds = append(strategy.crds, components.NewVirtualMachineInstanceCrd())
-	strategy.crds = append(strategy.crds, components.NewPresetCrd())
-	strategy.crds = append(strategy.crds, components.NewReplicaSetCrd())
-	strategy.crds = append(strategy.crds, components.NewVirtualMachineCrd())
-	strategy.crds = append(strategy.crds, components.NewVirtualMachineInstanceMigrationCrd())
+	strategy.crds = append(strategy.crds, patchValidation(components.NewVirtualMachineInstanceCrd(), "vmi"))
```
Comment (davidvossel) [inline pkg/virt-operator/install-strategy/strategy.go:241]: Rather than patching the CRDs here in the install-strategy logic, can we add the crd.Spec.Validation in the functions that create the CRDs structs themselves?

for example, for the VMI crd, instead of patching an already created CRD object, the validation could be added to the CRD in the components.NewVirtualMachineInstanceCrd() function.

File: tools/crd-validation-generator/validation-generator.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	for k, v := range *validations {
+		b, _ := yaml.Marshal(v)
```
Comment (davidvossel) [inline tools/crd-validation-generator/validation-generator.go:57]: since this is our own internal logic, why do the yaml conversion at all? Why don't we keep it in JSON instead? Won't that be more efficient?

**Comment by davidvossel** [comment]
> Once you have openapi-gen created you can do something like:
>cd staging/src/kubevirt.io/client-go/api/v1/
penapi-gen --alsologtostderr -i ./ -o ./staging/src/kubevirt.io/client-go/api/v1/ -p blub -h >~/go/src/kubevirt.io/kubevirt/hack/boilerplate/boilerplate.go.txt
and it will generate the openapi specification instead of our mechanism.

to be clear, this handles the openapi generation, but not the CRD's schema validation 

We need both.

I haven't found an existing way to generate the CRD schema accurately from the openapi logic... it's pretty frustrating. It appears in order to go with the flow with how the upstream ecosystem is generating the CRD validation, the controller-tools is the only approach. 

The existing tooling involves generating the openapi and crd validation schema independently apparently.

File: pkg/virt-operator/install-strategy/strategy.go (hunk 1 above)
Comment (xpivarc) [reply pkg/virt-operator/install-strategy/strategy.go]: Had some problem with validation so It would introduce some risk. This seems to be gone with the newer version controller-gen.

File: staging/src/kubevirt.io/client-go/Makefile (hunk 3 above)
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/Makefile]: Removed

File: pkg/virt-operator/install-strategy/validations_generated.go (hunk 2 above)
Comment (xpivarc) [reply pkg/virt-operator/install-strategy/validations_generated.go]: I am opened to any suggestions. What do you think about `CRDsValidation`?

File: tools/crd-validation-generator/validation-generator.go (hunk 4 above)
Comment (xpivarc) [reply tools/crd-validation-generator/validation-generator.go]: I am opened to any suggestions. What do you think about `CRDsValidation`?

File: pkg/virt-operator/install-strategy/strategy.go (hunk 5 above)
Comment (xpivarc) [reply pkg/virt-operator/install-strategy/strategy.go]: moved and tested

File: tools/crd-validation-generator/validation-generator.go (hunk 6 above)
Comment (xpivarc) [reply tools/crd-validation-generator/validation-generator.go]: I find `yaml` more readable, wdyt?

**Comment by xpivarc** [comment]
Uncover problem with `make generate` where our `Patch` in swagger is replaced by k8s one...

**Comment by xpivarc** [comment]
/cc @davidvossel @rmohr  
Should be ready for review

File: pkg/virt-operator/creation/components/crds.go [hunk 7]
```diff
@@ -37,6 +39,44 @@ const (
...
+	crd.Spec.PreserveUnknownFields = &PreserveUnknownFieldsFalse
+	validation, ok := CRDsValidation[crd.Spec.Names.Singular]
+	if !ok {
+		return
+	}
+	crvalidation := extv1beta1.CustomResourceValidation{}
+	k8syaml.NewYAMLToJSONDecoder(strings.NewReader(validation)).Decode(&crvalidation)
+	crd.Spec.Validation = &crvalidation
+	// workaround issues with metadata in spec
+	// TODO: find out if tests can be changed and this removed
```
Comment (davidvossel) [inline pkg/virt-operator/creation/components/crds.go:64]: I think I know what's going on here. For now, keep this code. I'll follow up this and include you

File: pkg/virt-operator/install-strategy/strategy.go [hunk 8]
```diff
@@ -175,7 +175,10 @@ func dumpInstallStrategyToBytes(strategy *InstallStrategy) []byte {
 		marshalutil.MarshallObject(entry, writer)
 	}
 	for _, entry := range strategy.crds {
-		marshalutil.MarshallObject(entry, writer)
+		b, _ := yaml.Marshal(entry)
+		writer.Write([]byte("---\n"))
+		writer.Write(b)
+		// marshalutil.MarshallObject(entry, writer)
```
Comment (davidvossel) [inline pkg/virt-operator/install-strategy/strategy.go:181]: this commented out code needs to be removed

File: pkg/virt-operator/creation/components/crds.go [hunk 9]
```diff
@@ -37,6 +39,44 @@ const (
...
+		spec := crd.Spec.Validation.OpenAPIV3Schema.Properties["spec"]
+		dataVolumeTemplates := spec.Properties["dataVolumeTemplates"]
+		items := dataVolumeTemplates.Items
+		metadata := items.Schema.Properties["metadata"]
+		metadata.Nullable = true
+		t := true
+		metadata.XPreserveUnknownFields = &t
+
+		items.Schema.Properties["metadata"] = metadata
+		// dataVolumeTemplates.Items = items
```
Comment (davidvossel) [inline pkg/virt-operator/creation/components/crds.go:75]: this commented out code should be removed

File: tools/crd-validation-generator/validation-generator.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	"strings"
+
+	"github.com/ghodss/yaml"
+	extv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func main() {
+
+	dirname := flag.String("crdDir", "./manifests/generated/", "path to directory with crds from where validation field will be parsed")
```
Comment (davidvossel) [inline tools/crd-validation-generator/validation-generator.go:18]: wouldn't `staging/src/kubevirt.io/client-go/config/crd/` be a better default at this point?

File: hack/generate.sh [hunk 11]
```diff
@@ -68,19 +68,43 @@ client-gen --clientset-name versioned \
...
+    cd config/crd
+    # tail -n +3 kubevirt.io_kubevirts.yaml > kubevirt.io_kubevirts_new.yaml
+    for file in *; do
+        tail -n +3 $file >$file"new"
+        mv $file"new" $file
+    done
+    cd ${KUBEVIRT_DIR}/tools/crd-validation-generator/ && go_build
+
+    cd ${KUBEVIRT_DIR}
+    ${KUBEVIRT_DIR}/tools/crd-validation-generator/crd-validation-generator -crdDir=${KUBEVIRT_DIR}/staging/src/kubevirt.io/client-go/config/crd/
```
Comment (davidvossel) [inline hack/generate.sh:89]: since we're setting the crdDir, would it make sense to explicitly set the output dir here as an argument as well?

**Review by davidvossel — COMMENTED** [review]
Looks great. i made a few comments. I think this is really close

File: pkg/virt-operator/creation/components/crds.go [hunk 12]
```diff
@@ -37,6 +39,44 @@ const (
...
+)
+
+func patchValidation(crd *extv1beta1.CustomResourceDefinition) {
+	crd.Spec.PreserveUnknownFields = &PreserveUnknownFieldsFalse
+	validation, ok := CRDsValidation[crd.Spec.Names.Singular]
+	if !ok {
+		return
+	}
+	crvalidation := extv1beta1.CustomResourceValidation{}
+	k8syaml.NewYAMLToJSONDecoder(strings.NewReader(validation)).Decode(&crvalidation)
```
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-operator/creation/components/crds.go:61]: I covered this in `crds_test.go`. Do you think we need to propagate it?

File: pkg/virt-operator/creation/components/crds.go (hunk 7 above)
Comment (xpivarc) [reply pkg/virt-operator/creation/components/crds.go:64]: The problem here is that we have no control over this API. So I couldn't use `// +kubebuilder:pruning:PreserveUnknownFields` as in `VirtualMachineInstanceTemplateSpec`.
I think recommended fix would be to replace a `metav1.ObjectMeta` with `TemplateObjectMeta`. Reasoning behind this is that this embedded metadata doesn't need all attributes of `metav1.ObjectMeta`. I want to have a look to Kubernetes codebase to better understand how validation is handled.

File: pkg/virt-operator/install-strategy/strategy.go (hunk 8 above)
Comment (xpivarc) [reply pkg/virt-operator/install-strategy/strategy.go:181]: Thanks

File: tools/crd-validation-generator/validation-generator.go (hunk 10 above)
Comment (xpivarc) [reply tools/crd-validation-generator/validation-generator.go:18]: indeed

**Comment by xpivarc** [comment]
/hold
Have an issue after rebase: `staging/src/kubevirt.io/client-go/api/v1/types.go:1443:25: unsupported type "float64"`

File: pkg/virt-operator/creation/components/crds.go (hunk 7 above)
Comment (davidvossel) [reply pkg/virt-operator/creation/components/crds.go:64]: with #4261 you should be able to add the perserve unknown fields marker now to DataVolumeTemplateSpec just like with VirtualMachineInstanceTemplateSpec.

**Comment by davidvossel** [comment]
> Have an issue after rebase: staging/src/kubevirt.io/client-go/api/v1/types.go:1443:25: unsupported type "float64"

when you're looking into this, can also revise the PR to take advantage of the changes I made to the DataVolumeTemplate api. I think you can remove that execption and add the preserve unknown fields to the api struct itself now after rebasing.

**Comment by xpivarc** [comment]
I reduced the scope to vm/vmi. I had to rebase several times to this point and new issues were introduced, making it harder for me to get this work... I will have follow up PR/s for other objects.

I discoverd few issues:
1. PR introducing configuration to KV have structs like:
`something int json:"something,string,omitempty"` which will generate wrong schema `type: integer.` I will send a PR to controller-gen.
2. `list_type_missing` is sometimes hard to resolve as you need to do the following:
a) mark the field with `//+listType=map`,  `//+listMapKey=somekey`
b) make sure that key can have default. Meaning it can't use `omitempty`. IMO this is example which can't be resolved https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2875

Note: We don't support SSA so 2. can be ignored.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go [hunk 13]
```diff
@@ -398,7 +397,8 @@ type Input struct {
 // +k8s:openapi-gen=true
 type Filesystem struct {
 	// Name is the device name
-	Name     string              `json:"name"`
+	Name string `json:"name"`
+	//
```
Comment (ashleyschuett) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:401]: I think this needs to be filled?

File: api/api-rule-violations-known.list [hunk 14]
```diff
@@ -152,6 +153,7 @@ API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,KubeVirtConfi
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,KubeVirtConfiguration,SupportedGuestAgentVersions
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,KubeVirtList,Items
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,KubeVirtStatus,Conditions
+API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,NodePlacement,Tolerations
```
Comment (ashleyschuett) [inline api/api-rule-violations-known.list:156]: should we add the list type annotation for this and Filesystems instead of adding them to the known violations list?

File: pkg/virt-operator/creation/components/crds.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ashleyschuett) [reply pkg/virt-operator/creation/components/crds.go:66]: I think this still needs to be updated to return an error?

**Review by ashleyschuett — CHANGES_REQUESTED** [review]

File: staging/src/kubevirt.io/client-go/api/v1/componentconfig.go [hunk 15]
```diff
@@ -34,11 +34,10 @@ type NodePlacement struct {
 	// These are additional tolerations other than default ones.
 	// +kubebuilder:validation:Optional
 	// +optional
-	//+listType=map
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/componentconfig.go:37]: why was this deleted?

File: pkg/virt-operator/creation/components/crds.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/virt-operator/creation/components/crds.go:66]: > I covered this in crds_test.go. Do you think we need to propagate it?

yes, unfortunately we need to propagate it. I don't know why this would ever occur, but if a function returns an error, we need to process it. Otherwise we run the risk of undefined behavior by silently ignoring the error.

**Review by davidvossel — APPROVED** [review]
/lgtm

**Comment by xpivarc** [comment]
@rmohr Thanks for pointing me to failing 1.19 lane.
@ashleyschuett @davidvossel  I just added the last commit to fix the failing 1.19lane.

File: tests/kubectl_test.go [hunk 16]
```diff
@@ -26,15 +26,19 @@ var _ = Describe("oc/kubectl integration", func() {
...
+		}
+		Expect(err).NotTo(HaveOccurred(), stderr)
 		Expect(output).To(ContainSubstring("apiVersion	<string>"))
 		Expect(output).To(ContainSubstring("kind	<string>"))
 		Expect(output).To(ContainSubstring("metadata	<Object>"))
 		Expect(output).To(ContainSubstring("spec	<Object>"))
 		Expect(output).To(ContainSubstring("status	<Object>"))
 	},
-		table.Entry("[rfe_id:3810]explain vm", "vm"),
+		table.FEntry("[rfe_id:3810]explain vm", "vm"),
```
Comment (rmohr) [inline tests/kubectl_test.go:41]: `Entry`

**Review by rmohr — CHANGES_REQUESTED** [review]
A debug oversight.

File: tests/kubectl_test.go (hunk 16 above)
Comment (xpivarc) [reply tests/kubectl_test.go:41]: Thanks! Seems like rebase is also needed.

**Comment by xpivarc** [comment]
Had to rebase.

**Review by davidvossel — APPROVED** [review]
/lgtm

**Comment by xpivarc** [comment]
:partying_face:
